zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+        /// buying from the pool or by using complete set operations. It's therefore convenient to
+        /// batch this function together with a `buy_complete_set` with `amount` as amount of
+        /// complete sets to buy.
+        ///
+        /// The operation is currently limited to binary and scalar markets.
+        ///
+        /// # Complexity
+        ///
+        /// `O(n)` where `n` is the number of outcomes in the specified market.
+        #[pallet::call_index(6)]
+        #[pallet::weight(T::WeightInfo::deploy_pool(spot_prices.len() as u32))]
+        #[transactional]
+        pub fn deploy_pool(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
+            spot_prices: Vec<BalanceOf<T>>,
+            swap_fee: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let asset_count = T::MarketCommons::market(&market_id)?.outcomes() as u32;
+            ensure!(spot_prices.len() == asset_count as usize, Error::<T>::IncorrectVecLen);
+            Self::do_deploy_pool(who, market_id, amount, spot_prices, swap_fee)?;
+            Ok(Some(T::WeightInfo::deploy_pool(asset_count)).into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        #[require_transactional]
+        fn do_buy(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            asset_out: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            min_amount_out: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure!(amount_in != Zero::zero(), Error::<T>::ZeroAmount);
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Active, Error::<T>::MarketNotActive);
+            Self::try_mutate_pool(&market_id, |pool| {
+                ensure!(pool.contains(&asset_out), Error::<T>::AssetNotFound);
+                // Defensive check (shouldn't ever happen)!
+                ensure!(
+                    pool.calculate_spot_price(asset_out)? <= MAX_SPOT_PRICE.saturated_into(),
+                    Error::<T>::Unexpected
+                );
+                ensure!(amount_in <= pool.calculate_max_amount_in(), Error::<T>::NumericalLimits);
+                T::AssetManager::transfer(pool.collateral, &who, &pool.account_id, amount_in)?;
+                let FeeDistribution {
+                    remaining: amount_in_minus_fees,
+                    swap_fees: swap_fee_amount,
+                    external_fees: external_fee_amount,
+                } = Self::distribute_fees(market_id, pool, amount_in)?;
+                let swap_amount_out =
+                    pool.calculate_swap_amount_out_for_buy(asset_out, amount_in_minus_fees)?;
+                let amount_out = swap_amount_out.saturating_add(amount_in_minus_fees);
+                ensure!(amount_out >= min_amount_out, Error::<T>::AmountOutBelowMin);
```

Comment:
It probably makes more sense to check what the user gets in the end, otherwise user (more likely the frontend) always has to consider the (probably varying) fees when specifying this parameter. In the end, the user only wants to ensure that he actually receives a specific amount.

```suggestion
                ensure!(amount_out >= swap_amount_out, Error::<T>::AmountOutBelowMin);
```

Reply:
Not quite sure if I follow. Your suggestion doesn't work because it always succeeds, since `amount_out = swap_amount_out + amount_in_minus_fees`. I agrere that the user cares most about what ends up in their wallet. But that's exactly `amount_out`.

There may be a bit of confusion about what `amount_out` is. I agree that the code slightly obfuscated what's actually going on. Each buy consists of two steps: 1) Buy x complete sets. 2) Sell `x` of each outcome you don't want for more of the outcome you do want. Let's say you want to spend `x` dollars to buy token `i`. You first spend `x` dollars and buy `x` complete sets (first step) and then you sell `x` units of each token `j != i` for more units of `i` (step two). `y(x)` is the amount of tokens you get in step two. b is the liquidity parameter of the pool, `x` is the amount of the base asset the user is paying and `r_i` is the balance of outcome `i` in the pool. So to calculate the total amount of `i` you receive, you actually have to calculate `z(x) = y(x) + x`.

The actual implementation takes a shortcut by letting the pool buy the complete sets to avoid unnecessary storage reads/writes on the trader's account.

Reply:
Thanks for elaborating this. I misread the (accurate) docs :man_facepalming: 
> Note that the total amount of outcome i that Alice receives is y(x) from the O2O trade
and x from the complete set operation. We denote this by z(x) = y(x) + x.